src/models/document/base-document-content.ts

```diff
@@ -1077,5 +1064,148 @@ export const BaseDocumentContentModel = types
           // ignore shared models with errors
         }
       });
+    },
+    duplicateTiles(tiles: IDragTileItem[]) {
+      // Determine the row to add the duplicated tiles into
+      const rowIndex = self.getRowAfterTiles(tiles);
+
+      // Find shared models used by tiles being duplicated
+      const sharedModelEntries = Object.values(self.getSharedModelsUsedByTiles(tiles.map(tile => tile.tileId)));
+
+      // Update shared models with new ids
+      interface UpdatedSharedDataSet {
+        sharedModelId: string;
+        attributeIdMap: Record<string, string>;
+        caseIdMap: Record<string, string>;
+      }
+      const updatedSharedModelMap: Record<string, UpdatedSharedDataSet> = {};
+      const newSharedModelEntries: any[] = [];
+      sharedModelEntries.forEach(sharedModelEntry => {
+        // For now, only duplicate shared data sets
+        if (sharedModelEntry.sharedModel.type === "SharedDataSet") {
+          // Determine new ids
+          const sharedDataSet = getSnapshot(sharedModelEntry.sharedModel) as SharedDataSetSnapshotType;
+          const sharedModelId = uniqueId();
+          const dataSetId = uniqueId();
+          const attributeIdMap: Record<string, string> = {};
+          const caseIdMap: Record<string, string> = {};
+          sharedDataSet.dataSet.attributes?.forEach((attr: any) => {
+            attributeIdMap[attr.id] = uniqueId();
+          });
+          sharedDataSet.dataSet.cases?.forEach((c: any) => {
+            caseIdMap[c.__id__] = newCaseId();
+          });
+          if (sharedDataSet.id) updatedSharedModelMap[sharedDataSet.id] = { sharedModelId, attributeIdMap, caseIdMap };
+
+          // Create a snapshot for the shared model with updated ids, which will be updated with new tile ids
+          // and added to the document later
+          const newAttributes = sharedDataSet.dataSet.attributes?.map((a: any) => {
+            return { ...a, id: attributeIdMap[a.id] };
+          });
+          const newCases = sharedDataSet.dataSet.cases?.map((c: any) => ({ ...c, __id__: caseIdMap[c.__id__] }));
+          newSharedModelEntries.push({
+            tiles: sharedModelEntry.tiles,
+            sharedModel: {
+              ...sharedDataSet,
+              id: sharedModelId,
+              dataSet: {
+                ...sharedDataSet.dataSet,
+                id: dataSetId,
+                attributes: newAttributes,
+                cases: newCases
+              }
+            }
+          });
+        }
+      });
+
+      // Update tile content with new shared model ids
+      const tileIdMap: Record<string, string> = {};
+      const updatedTiles: any[] = [];
+      tiles.forEach(tile => {
+        const oldContent = JSON.parse(tile.tileContent);
+        const tileContent = oldContent;
+        tileIdMap[tile.tileId] = tile.newTileId ?? uniqueId();
```

Comment:
I think you probably want to create a new uniqueId in all cases. See my other comment.